routers/api/v1/repo/commits.go

```diff
@@ -120,3 +122,204 @@ func GetSingleCommit(ctx *context.APIContext) {
+	if err != nil {
+		ctx.ServerError("OpenRepository", err)
+		return
+	}
+
+	page := ctx.QueryInt("page")
+	if page <= 0 {
+		page = 1
+	}
+
+	sha := ctx.Query("sha")
+
+	var baseCommit *git.Commit
+	if len(sha) == 0 {
+		// no sha supplied - use default branch
+		head, err := gitRepo.GetHEADBranch()
+		if err != nil {
+			ctx.ServerError("GetHEADBranch", err)
+			return
+		}
+		baseCommit, err = gitRepo.GetBranchCommit(head.Name)
+		if err != nil {
+			ctx.ServerError("GetCommit", err)
+			return
+		}
+	} else {
+		// get commit specified by sha
+		baseCommit, err = gitRepo.GetCommit(sha)
+		if err != nil {
+			ctx.ServerError("GetCommit", err)
+			return
+		}
+	}
+
+	// Total commit count
+	commitsCountTotal, err := baseCommit.CommitsCount()
+	if err != nil {
+		ctx.ServerError("GetCommitsCount", err)
+		return
+	}
+
+	pageCount := int(math.Ceil(float64(commitsCountTotal) / float64(git.CommitsRangeSize)))
+
+	// Query commits
+	commits, err := baseCommit.CommitsByRange(page)
+	if err != nil {
+		ctx.ServerError("CommitsByRange", err)
+		return
+	}
+
+	userCache := make(map[string]*models.User)
+
+	apiCommits := make([]*api.Commit, commits.Len())
+
+	i := 0
+	for commitPointer := commits.Front(); commitPointer != nil; commitPointer = commitPointer.Next() {
+		commit := commitPointer.Value.(*git.Commit)
+
+		var apiAuthor, apiCommitter *api.User
+
+		// Retrieve author and committer information
+		cacheAuthor, ok := userCache[commit.Author.Email]
+		if ok {
+			apiAuthor = cacheAuthor.APIFormat()
+		} else {
+			author, err := models.GetUserByEmail(commit.Author.Email)
+			if err != nil && !models.IsErrUserNotExist(err) {
+				ctx.ServerError("Get user by author email", err)
+				return
+			} else if err == nil {
+				apiAuthor = author.APIFormat()
+				userCache[commit.Author.Email] = author
+			}
+		}
+		cacheCommitter, ok := userCache[commit.Committer.Email]
+		if ok {
+			apiCommitter = cacheCommitter.APIFormat()
+		} else {
+			committer, err := models.GetUserByEmail(commit.Committer.Email)
+			if err != nil && !models.IsErrUserNotExist(err) {
+				ctx.ServerError("Get user by committer email", err)
+				return
+			} else if err == nil {
+				apiCommitter = committer.APIFormat()
+				userCache[commit.Committer.Email] = committer
+			}
+		}
+
+		// Retrieve parent(s) of the commit
+		apiParents := make([]*api.CommitMeta, commit.ParentCount())
+		for i := 0; i < commit.ParentCount(); i++ {
+			sha, _ := commit.ParentID(i)
+			apiParents[i] = &api.CommitMeta{
+				URL: ctx.Repo.Repository.APIURL() + "/git/commits/" + sha.String(),
+				SHA: sha.String(),
+			}
+		}
+
+		// Create json struct
+		apiCommits[i] = &api.Commit{
+			CommitMeta: &api.CommitMeta{
+				URL: ctx.Repo.Repository.APIURL() + "/git/commits/" + commit.ID.String(),
+				SHA: commit.ID.String(),
+			},
+			HTMLURL: ctx.Repo.Repository.HTMLURL() + "/commits/" + commit.ID.String(),
+			RepoCommit: &api.RepoCommit{
+				URL: setting.AppURL + ctx.Link[1:],
```

Comment:
This should be something like `"http://localhost:3000/api/v1/repos/root/test/git/commits/3f312aa93b50477c0d23e1c27b2563d4025f466d",` yet I'm getting it without a SHA: `"http://localhost:3000/api/v1/repos/root/test/git/commits/",`

Reply:
fixed